server/bytestream/bytestream.go

```diff
@@ -14,11 +15,140 @@ import (
 	"github.com/buildbuddy-io/buildbuddy/server/util/status"
 	"google.golang.org/protobuf/proto"
 
+	arpb "github.com/buildbuddy-io/buildbuddy/proto/archive"
 	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
 	bspb "google.golang.org/genproto/googleapis/bytestream"
 )
 
+func FetchBytestreamZipManifest(ctx context.Context, env environment.Env, url *url.URL) (*arpb.ArchiveManifest, error) {
+	r, err := digest.ParseDownloadResourceName(strings.TrimPrefix(url.RequestURI(), "/"))
+	if err != nil {
+		return nil, err
+	}
+	// Let's just read 64K and see if we can find the central directory in there.
+	// We probably don't want to be in the business of rendering 3000-file zips'
+	// contents anyway.
+	offset := r.GetDigest().GetSizeBytes() - 65536
+	if offset < 0 {
+		offset = 0
+	}
+
+	reader, writer := io.Pipe()
+	go func() {
+		err = StreamBytestreamFileChunk(ctx, env, url, offset, r.GetDigest().GetSizeBytes()-offset, func(data []byte) {
+			writer.Write(data)
+		})
+		writer.CloseWithError(err)
+	}()
+
+	// Dump the full contents out into a buffer (should be 64K or less).
+	footer, err := io.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	// Find and parse the End of Central Directory header or fail.
+	eocd, err := readDirectoryEnd(footer, r.GetDigest().GetSizeBytes())
+	if err != nil {
+		return nil, err
+	}
```

Comment:
instead of using a pipe and a goroutine, you can use bytes.Buffer (which basically just implements io.Reader and io.Writer interfaces on a []byte) and stream your bytes into the buffer then read from it.

Something like:
```
var buf bytes.Buffer
err = StreamBytestreamFileChunk(ctx, env, url, offset, r.GetDigest().GetSizeBytes()-offset, func(data []byte) {
  buf.Write(data)
})
if err != nil {
  return nil, err
}
eocd, err := readDirectoryEnd(buf.Bytes(), r.GetDigest().GetSizeBytes())
```

Reply:
Thanks for this, it's a lot less ugly for the cases where we know the (small) amount of data we need.

I updated StreamByteStreamFile to take a Writer, so that meant updating the little api_server call to provide a writer, I'm not sure if I've quite done it in the "idiomatic" way, so it'd be good if you could take a peek at that.